src/appleseed-max-impl/appleseedrenderer/dialoglogtarget.cpp

```diff
@@ -0,0 +1,259 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "dialoglogtarget.h"
+
+// appleseed-max headers.
+#include "appleseedrenderer/resource.h"
+#include "main.h"
+#include "utilities.h"
+
+// appleseed.foundation headers.
+#include "foundation/platform/windows.h"    // include before 3ds Max headers
+#include "foundation/utility/string.h"
+
+// Boost headers.
+#include "boost/thread/locks.hpp"
+#include "boost/thread/mutex.hpp"
+
+// 3ds Max headers.
+#include <max.h>
+
+// Standard headers.
+#include <Richedit.h>
+
+namespace asf = foundation;
+namespace asr = renderer;
+
+namespace
+{
+    boost::mutex                g_message_queue_mutex;
+    std::vector<MessagePair>    g_message_queue;
+    HWND                        g_log_dialog = 0;
+
+    static INT_PTR CALLBACK dialog_proc(HWND hwnd, UINT msg, WPARAM wparam, LPARAM lparam)
+    {
+        switch (msg)
+        {
+          case WM_INITDIALOG:
+            break;
+          case WM_CLOSE:
+          case WM_DESTROY:
+            if (g_log_dialog != 0)
+            {
+                GetCOREInterface14()->UnRegisterModelessRenderWindow(g_log_dialog);
+                g_log_dialog = 0;
+                DestroyWindow(hwnd);
+            }
+            break;
+          
+          case WM_SIZE:
+          {
+            HWND edit_box = GetDlgItem(hwnd, IDC_EDIT_LOG);
+            MoveWindow(
+                edit_box,
+                0,
+                0,
+                LOWORD(lparam),
+                HIWORD(lparam),
+                true);
+            return TRUE;
+          }
+            break;
+
+          default:
+            return FALSE;
+        }
+        return TRUE;
+    }
+
+    void append_text(HWND edit_box, const wchar_t* text, COLORREF color)
+    {
+        const int text_length = GetWindowTextLength(edit_box);
+        SendMessage(edit_box, EM_SETSEL, text_length, text_length);
+
+        CHARFORMAT char_format;
+        SendMessage(edit_box, EM_GETCHARFORMAT, SCF_SELECTION, reinterpret_cast<LPARAM>(&char_format));
+        char_format.cbSize = sizeof(char_format);
+        char_format.dwMask = CFM_COLOR | CFM_FACE;
+        char_format.dwEffects = 0;
+        char_format.crTextColor = color;
+        wcscpy(reinterpret_cast<WCHAR*>(char_format.szFaceName), L"Consolas");
```

Comment:
Is the `reinterpret_cast<>` required here? `CHARFORMAT.szFaceName` appears to be of type `TCHAR` of which the [documentation](https://msdn.microsoft.com/en-us/library/office/cc842072.aspx) says:

> For Unicode platforms, TCHAR is defined as synonymous with the WCHAR type.

Reply:
Thanks! This part was very confusing.